src/railfence.rs

```diff
@@ -0,0 +1,284 @@
+//! The Railfence Cipher is a transposition cipher. It has a very low keyspace and is therefore
+//!incredibly insecure.
+//!
+//! This implementation currently transposes all input characters including whitespace and
+//!punctuation.
+
+/// A Railfence cipher.
+///
+/// This struct is created by the `new()` method. See its documentation for more.
+pub struct Railfence {
+    key: usize,
+}
+
+impl Railfence {
+    /// Initialise a Railfence cipher given a specific key.
+    ///
+    /// Will return `Err` if the key is zero.
+    pub fn new(key: usize) -> Result<Railfence, &'static str> {
+        if key == 0 {
+            return Err("Invalid key. Railfence key cannot be zero.");
+        }
+
+        Ok(Railfence { key: key })
+    }
+
+    /// Encrypt a message using a Railfence cipher.
+    ///
+    /// # Examples
+    /// Basic usage:
+    ///
+    /// ```
+    /// use cipher_crypt::railfence::Railfence;
+    ///
+    /// let r = Railfence::new(3).unwrap();
+    /// assert_eq!("Src s!ue-ertmsaepseeg", r.encrypt("Super-secret message!"));
+    /// ```
+    pub fn encrypt(&self, message: &str) -> String {
+        // Encryption process:
+        //   First a table is created with a height given by the key and a length
+        //   given by the message length.
+        //   e.g.
+        //   For a key of 3 and the message "Hello, World!" of length 13:
+        //      .............
+        //      .............
+        //      .............
+        //   The message can then be written onto the grid in a zigzag going right:
+        //      H...o...o...!
+        //      .e.l.,.W.r.d.
+        //      ..l... ...l..
+        //   The encrypted message is then read line by line:
+        //      Hoo!el,Wrdl l 
+        
+        
+        // A key of one causes a problem when calculating the cycle later on,
+        // so just return the encrypted message now
+        if self.key == 1 {
+            return message.to_string()
+        }
+
+        // Create the table that will be used for encryption
+        // The form of an entry is (bool, char).
+        // The bool determines whether the current entry is being used, and if so
+        // the char is part of the plain/cipher text
+        let mut table = vec![vec![(false, 'a'); message.len()]; self.key];
+
+        for (i, c) in message.chars().enumerate() {
+            let col = i;
+            // In the railfence cipher the letters are placed diagonally in a zigzag,
+            // so, with a key of 4 say, the row numbers will go
+            //      0, 1, 2, 3, 2, 1, 0, 1, 2, 3, 2, 1, 0, ...
+            // This repeats with a cycle (or period) given by (2*key - 2)
+            //      [0, 1, 2, 3, 2, 1], [0, 1, 2, 3, 2, 1], 0, ...
+            // This cycle is always even.
+            let cycle = 2 * self.key - 2;
```

Comment:
Code between lines 74 - 83, could be moved into its own private function as its used both here and in the decrypt function.